### ISSUE TITLE
Make sure we are recording the likelihood from just the evaluation set

### ISSUE THREAD

**Issue by alexpiet** [AllenNeuralDynamics/aind-disrnn-wrapper#5]
# Summary

The dimensions of yhat, and network state are the size of the full dataset, not the evaluation set. This means that the `likelihood` recorded in the outputs.json is probably the likelihood on the full dataset, and not on the evaluation set. This should NOT impact the training/evaluation loss (includes the penalties) reported by `rnn_utils.train_network()`

## details
This is a problem inherited from my code that I haven't gotten a round to investigating. Documenting here since other people are now using

In this [code](https://github.com/AllenNeuralDynamics/aind-disrnn-wrapper/blob/12b01428b519e9f3fecdf3cd456e89490aca6d40/code/run_capsule.py#L282)
```python
    xs, ys = next(dataset_eval)
    yhat, network_states = rnn_utils.eval_network(
        lambda: disrnn.HkDisentangledRNN(noiseless_network), params, xs
    )
```
we grab the evaluation dataset, evaluate the network, and get the yhat and network states.

Later we use those variables to compute the likelihood. 
```python
    logger.info("Output details:")
    logger.info(asset_name)
    logger.info("yhat size: {}".format(np.shape(yhat)))
    logger.info("state size: {}".format(np.shape(network_states)))

    likelihood = rnn_utils.normalized_likelihood(ys, yhat[:, :, 0:2])
    output["likelihood"] = likelihood
```

But! I've noticed that the dimensions of `yhat` and `network_states` are always for the FULL dataset, not just the evaluation dataset. In this example run

```
INFO:2025-11-14 20-28-12:run_capsule.py:72:    Dataset details:
INFO:2025-11-14 20-28-12:run_capsule.py:73:    Input size: (580, 31, 2)
INFO:2025-11-14 20-28-12:run_capsule.py:74:    Output size: (580, 31, 1)
INFO:2025-11-14 20-28-12:run_capsule.py:75:    Train output: (580, 16, 1)
INFO:2025-11-14 20-28-12:run_capsule.py:76:    Test output: (580, 15, 1)
...
INFO:2025-11-14 20-58-48:run_capsule.py:206:    Output details:
INFO:2025-11-14 20-58-48:run_capsule.py:207:    /data/disrnn_dataset_774212/disrnn_dataset.csv
INFO:2025-11-14 20-58-48:run_capsule.py:208:    yhat size: (580, 31, 3)
INFO:2025-11-14 20-58-48:run_capsule.py:209:    state size: (580, 31, 5) 
```

## to resolve
- [ ] understand what is going wrong
- [ ] record likelihood from just the evaluation set
- [ ] record latents from full dataset, useful for analysis

**Comment by alexpiet** [comment]
@rachelstephlee doesn't see this in the gru_trainer. Notably it uses this line: 
```python
xs_eval, ys_eval = dataset_eval.get_all()
```
whereas the disrnn trainer has 
```python
xs, ys = next(dataset_eval)
```

Should be a simple fix to update how we are calling the evaluation dataset.

**Comment by alexpiet** [comment]
It might be the case that the "next()" built in operator doesn't work with rnn_utils.split_dataset()

to resolve
- [ ] update our code to use `get_all()`
- [ ] investigate and document whether `next()` is not working with `rnn_utils()`

**Comment by hanhou** [comment]
### The original issue
The original problem has been fixed by Kevin's [PR](https://github.com/AllenNeuralDynamics/aind-disentangled-rnns/pull/6/changes). 

The root cause is when we split the dataset ([code](https://github.com/AllenNeuralDynamics/aind-disentangled-rnns/blob/d9c0a7e33c828cb4e682616ec17b7a26f70b9684/disentangled_rnns/library/rnn_utils.py#L280)), we set `dataset_train.batch_size = {the batch_size of the **_full_** dataset}`:
```python
  dataset_eval = DatasetRNN(
      xs[:, eval_sessions, :],
      ys[:, eval_sessions, :],
      x_names=dataset.x_names,
      y_names=dataset.y_names,
      y_type=dataset.y_type,
      n_classes=dataset.n_classes,
      batch_size=dataset.batch_size,   # --> use the _full_ dataset batch_size as the eval batch_size
      batch_mode=dataset.batch_mode,
  )
```
Afterwards, when `next()` is called ([code](https://github.com/AllenNeuralDynamics/aind-disentangled-rnns/blob/d9c0a7e33c828cb4e682616ec17b7a26f70b9684/disentangled_rnns/library/rnn_utils.py#L209-L244)), if `batch_mode = "random"` (by [default](https://github.com/AllenNeuralDynamics/aind_disrnn_utils/blob/94a1b1f23c1bed67983424d188d2bc89b10f4edb/src/aind_disrnn_utils/data_loader.py#L101)), eval sessions are sampled with replacement, so that the number of sampled sessions matches the _**full**_ dataset, not the eval dataset.
```python
    elif self.batch_mode == 'random':
      inds_to_get = np.random.choice(self._n_episodes, size=self.batch_size)   # --> select 31 sessions out of 15 eval sessions
      return self._xs[:, inds_to_get], self._ys[:, inds_to_get]
```
In other words, the original likelihood _**does**_ come from eval dataset, although some eval sessions are used more than once.

### Current status
In the latest [code](https://github.com/AllenNeuralDynamics/aind-disentangled-rnns/blob/6842bdc06d3f2135be368f9d28b42da717b1ee73/disentangled_rnns/library/rnn_utils.py#L274-L299):

```python
  if dataset.batch_mode == 'single':
    batch_size = None
  else:
    batch_size = dataset.batch_size

  dataset_train = DatasetRNN(
      xs[:, train_sessions, :],
      ys[:, train_sessions, :],
      x_names=dataset.x_names,
      y_names=dataset.y_names,
      y_type=dataset.y_type,
      n_classes=dataset.n_classes,
      batch_size=batch_size,    # --> POTENTIAL BUG: if dataset.batch_mode != 'single', batch_size > len(train_sessions)
      batch_mode=dataset.batch_mode,
  )
  dataset_eval = DatasetRNN(
      xs[:, eval_sessions, :],
      ys[:, eval_sessions, :],
      x_names=dataset.x_names,
      y_names=dataset.y_names,
      y_type=dataset.y_type,
      n_classes=dataset.n_classes,
      batch_size=None,     # --> the batch_size will always be len(eval_sessions), thus fixing the original bug
      batch_mode=dataset.batch_mode,
  )
```
The `dataset_eval` has the `batch_size=len(eval_sessions)`, fixing the original bug. However, there are two caveats:

1. If we still use `xs_eval, ys_eval = next(dataset_eval)` with `batch_mode="random"`, the length of `xs_eval` now equals `len(eval_sessions)`, but their contents are not exactly the same due to the random sample step. In other words, we should always use `.get_all()` for evaluation (like in the [PR](https://github.com/AllenNeuralDynamics/aind-disrnn-wrapper/pull/31)). 
2. When `batch_mode != "single"`, the batch size of `dataset_train` is still the batch size of `dataset`, meaning that, at each step of training, `len(full_dataset)` number of sessions are randomly sampled from `len(train_dataset)` sessions. See [code](https://github.com/AllenNeuralDynamics/aind-disentangled-rnns/blob/6842bdc06d3f2135be368f9d28b42da717b1ee73/disentangled_rnns/library/rnn_utils.py#L781-L787)
```python
# If the training dataset uses batching, get a new batch
if training_dataset.batch_mode != 'single':
  xs_train, ys_train = next(training_dataset)

loss, params, opt_state = train_step(
    params, opt_state, xs_train, ys_train, subkey_train
)
```

As confirmed by this run
```python
# Mouse 779531, 22 sessions in total
xs_train.shape
# (616, 22, 2)
training_dataset._xs.shape
# (616, 11, 2)
```

This will not introduce any severe issue like data leaking, but it would be great to fix.

**Comment by hanhou** [comment]
Related to Kevin's latest message:

<img width="722" height="286" alt="Image" src="https://github.com/user-attachments/assets/df3de130-2ecb-4f61-abb2-4340ba13437f" />

Having (much) larger batch size than the entire dataset is fine, and it is a good practice to explicitly specify it when creating the dataset. In fact, he suggested using 1024.
